File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p03825/p03825_num4_parsed.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 1000000007;

int main() {
    int N, M, K;
    cin >> N >> M >> K;

    vector<vector<int>> dp(N + 1, vector<int>(M + 1, 0));
    dp[0][0] = 1;

    for (int i = 1; i <= N; i++) {
        for (int j = 0; j <= M; j++) {
            for (int k = 1; k <= K - 1 && k <= i; k++) {
                dp[i][j] = (dp[i][j] + dp[i - k][j + 1]) % MOD;
            }
        }
    }

    for (int j = 1; j <= M; j++) {
        for (int k = 1; k <= K - 1 && k <= N; k++) {
            dp[N][j] = (dp[N][j] + dp[N - k][j - 1]) % MOD;
        }
    }

    int result = 0;
    for (int j = 0; j <= M; j++) {
        result = (result + dp[N][j]) % MOD;
    }

    cout << result << endl;

    return 0;
}

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p03825/p03825_num4_parsed.py
MOD = 10**9 + 7

def calculate_combinations(n, r):
    result = 1
    for i in range(1, r + 1):
        result = (result * (n - i + 1) // i) % MOD
    return result

def calculate_possible_values(n, m, k):
    result = 0
    for i in range(1, k + 1):
        if (n + m - 1) % (k - 1) == 0:
            combinations = calculate_combinations(n + m - 1, i)
            result = (result + combinations) % MOD
    return result

# Read input
n, m, k = map(int, input().split())

# Calculate and print the number of possible values
print(calculate_possible_values(n, m, k))